lms/djangoapps/course_blocks/transformers/library_content.py

```diff
@@ -176,3 +178,48 @@ def publish_event(event_name, result, **kwargs):
             format_block_keys,
             publish_event,
         )
+
+
+class ContentLibraryOrderTransformer(BlockStructureTransformer):
+    """
+    A transformer that manipulates the block structure by modifying the order of the
+    selected blocks within a library_content module to match the order of the selections
+    made by the ContentLibraryTransformer or the corresponding XBlock. So this transformer
+    requires the selections for the randomized content block to be already
+    made either by the ContentLibraryTransformer or the XBlock.
+
+    Staff users are *not* exempted from library content pathways/
+    """
+    WRITE_VERSION = 1
+    READ_VERSION = 1
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class;
+        same identifier used in setup.py
+        """
+        return "library_content_randomize"
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Collects any information that's necessary to execute this
+        transformer's transform method.
+        """
+        # There is nothing to collect.
+        pass
+
+    def transform(self, usage_info, block_structure):
+        """
+        Transforms the order of the children of the randomized content block
+        to match the order of the selections made and stored in the XBlock 'selected' field.
+        """
+        for block_key in block_structure:
+            if block_key.block_type != 'library_content':
+                continue
+
+            state_dict = get_student_module_as_dict(usage_info.user, usage_info.course_key, block_key)
+            library_children = block_structure.get_children(block_key)
+            ordering_data = {block[1]: position for position, block in enumerate(state_dict['selected'])}
+            library_children.sort(key=lambda block, data=ordering_data: data[block.block_id])
```

Comment:
`lms.djangoapps.grades.tasks.recalculate_subsection_grade_v3` was failing on this line with the following stack trace:

```
exceptions:KeyError: u'23e9f816a170b4eb6422'
Traceback (most recent call last):
File "/edx/app/edxapp/edx-platform/manage.py", line 123, in <module>
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/djcelery/management/commands/celery.py", line 21, in run_from_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 793, in execute_from_commandline
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/base.py", line 311, in execute_from_commandline
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 785, in handle_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 717, in execute
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/worker.py", line 179, in run_from_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/base.py", line 274, in __call__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/worker.py", line 212, in run
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/worker/__init__.py", line 206, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bootsteps.py", line 123, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bootsteps.py", line 374, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/base.py", line 131, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/prefork.py", line 119, in on_start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/asynpool.py", line 401, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 972, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/asynpool.py", line 415, in _create_worker_process
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 1068, in _create_worker_process
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/process.py", line 137, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/forking.py", line 105, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/process.py", line 292, in _bootstrap
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 295, in run
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 367, in workloop
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 349, in _fast_trace_task
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/application_celery.py", line 85, in wrapper
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 179, in recalculate_subsection_grade_v3
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 248, in _recalculate_subsection_grade
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 684, in retry
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/utils/__init__.py", line 248, in maybe_reraise
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 239, in _recalculate_subsection_grade
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 310, in _update_subsection_grades
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/api.py", line 97, in get_course_blocks
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/manager.py", line 76, in get_transformed
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/transformers.py", line 116, in transform
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/transformers.py", line 153, in _transform_without_filters
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/transformers/library_content.py", line 225, in transform
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/transformers/library_content.py", line 225, in <lambda>
```

Reply:
@lgp171188 ^

Reply:
@ormsbee, sorry for causing this issue. I will check this and try to reproduce it in my devstack.

This PR introduces a new `BlockStructureTransformer` sub-class `ContentLibraryOrderTransformer` to be run immediately after the `ContentLibraryTransformer`. The latter is a `FilteringTransformerMixin` sub-class and can only filter the children of a randomized content block and cannot randomize the order of selection.

This PR adds a change to the `ContentLibraryTransformer` to randomize/shuffle the order of the selected blocks before storing it in the `StudentModule`. The existing behaviour of making fresh selections only when the type of blocks to be chosen or the number of blocks to be chose has changed is retained as-is.

The new `ContentLibraryOrderTransformer` then tries arrange the children blocks based on the stored order using the sort method above which is throwing an exception in the task.

On taking a quick look at the exception and the code, it appears that this can happen only when the selected blocks stored in the `StudentModule` no longer match the ones that are actually available in the current block structure. And this appears to be a weird and unexpected error because the `ContentLibraryTransformer`, which makes the selections and stores it in the `UserModule` and the `ContentLibraryOrderTransformer` which uses it, are always run together one after the other.

Do you see any obvious issues with the approach or have any pointers on what might have gone wrong? That will be helpful because my understanding of the block structure and its transformation is limited and I could have missed something crucial while working on this.

That said, I will still try to debug and investigate this and come up with a fix. Thanks!

Reply:
It might be helpful to debug this issue further if there are some more details about the course/course structure are available so that I can try to reproduce the issue.